### PR TITLE
Shortcut to set current user as campaign manager

### LINKED ISSUE #417: Shortcut to set user as campaign manager
When creating a campaign, the user can search for a person in the organization to set as the campaign manager. Oftentimes the person creating the campaign will also be the manager, so it would be nice to have a shortcut to set the user themselves as the manager.

I imagine a little link under the search field in the create campaign form.

### PR DESCRIPTION
**Pull request by richardolsson**
## Changes in summary
This PR adds a shortcut under the "Campaign manager" field in the campaign form, to directly select the current user as the manager.

The shortcut is only visible when the user is not already the selected manager, i.e. when there is no manager assigned or when another person is assigned as manager.

To solve this, I had to switch the `Autocomplete` used to select manager to [controlled](https://reactjs.org/docs/forms.html#controlled-components). I think this is a fine solution, but I'm curious to hear the thoughts of others!

Resolves #417

## How to test
1. Navigate to the "All campaigns" page
2. Via the Speed Dial in the bottom right, click to create a new campaign
3. In the "Create new campaign" form, click the new "Select yourself as manager" – "Angela Davis" (or whatever your user is) should become selected and the link disappear
4. Search to select someone else – the link be vissible
5. Use the "x" to select no manager – the link be visible

## Screenshots
image

### CODE REVIEW

**Comment by djbusstop** [comment]
@richardolsson I reworked your PR for two reasons: 
1. I had massively changed the `CampaignDetailsForm` (previously `CampaignForm`) component in my PR re: form error handling, and merging this together didn't work, since I had replaced the file. It needed to be done in any case.
2. There were some things about the way searching was done that always bothered me, mostly regarding the naming, but also how there was a lot of logic to handle the searching for people feature that was just chucked in to the component. The way you solved the problem worked great, but I didn't love how that component was structured already and was thinking about using hooks to encapsulate the code anyway. That's why I created a hook for the `useSelectManager` which handles all this stuff.
  a. I think the `campaign.manager` type just taking an `id` and a `name` while most if not all other `person` (user, person, etc) models use `first_name` and `last_name` creates a nightmare for types and mixing of `Person` and `manager` types. I think the reworking I did is a bit more clear, at least to me, when we are using a `name` and `first_name`/`last_name`. The hook is always using the shape that the `campaign.manager` will return, where the autoselect always uses a subset of the the person data shape

Here's an overview of the chang

File: src/components/forms/CampaignDetailsForm/useSelectManager.tsx [hunk 1]
```diff
@@ -0,0 +1,47 @@
...
+    const [selectedManager, setSelectedManager] = useState(initialManager);
+    const user = useUser();
+
+    return {
+        managerSearchQuery,
+        managerSearchValue,
+        selectedManager,
+        setManagerSearchValue,
+        setSelectedManager,
+        setSelectedManagerToUser: () => user && setSelectedManager({ id: user.id, name: `${user.first_name} ${user.last_name}` }),
```
Comment (richardolsson) [inline src/components/forms/CampaignDetailsForm/useSelectManager.tsx:42]: If I'm not mistaken, there is a bug here, that did not exist in my original code.

This line of code is setting the manager to the `user` object, but the manager is not a _user_, it's a _person_. The `Person` object in Zetkin is not the same as `User`, and setting the `manager_id` to `user.id` (which is the end result of this) will result in another person (with the ID of the user's `User` object) being added as the manager.

This is what the `getUserMemberships()` function was for. It retrieves the memberships of the current user, in which the `Person` object for this specific organization can be found. It seems like this function is no longer being used here.

The reason why this issue does not become apparent with the dummy data, is that the ID of the `testadmin@example.com` user (2) is the same as the `Person` it's connected to in organization 1.

**Review by richardolsson — COMMENTED** [review]
I appreciate the initiative and the intent of these changes, but I think I disagree with the execution.

First, I believe that simplifying things this way has removed a significant piece of required logic and introduced a bug (see comment below).

But the other issue is that I think it's [premature abstraction](https://www.youtube.com/watch?v=wuVy7rwkCfc). By moving this logic out into a hook, we've abstracted it at what I believe will turn out to be the wrong level.

Finding and selecting a person is a very common thing in Zetkin, and I believe we will see it in many more places quite soon. When we do, we will want to use a single component for all of those cases, e.g. a `<PersonAutocomplete/>` or something.

I think we should have waited with this abstraction until we reach that point. It doesn't mean this hook is a bad thing, just that there is likely some amount of effort that went into it that will turn out to have been wasted when the real need for abstraction (DRY) appears.

> One other thing I noticed you had set the search results to include the current user as the top option every time the user searches. I worked to reimplement this feature and could still leave it in. But I don't think it's very necessary, since the "set yourself as manager" feature already does that for you, and it may be confusing when you search a name and your own name comes up first. What do you think?

I don't believe this understanding of my implementation is correct, but I may have made a mistake. What I intended, was for the _current selection_ to always be part of the `results` array passed to the `options` property, regardless of whether the current selection is the user's profile or not.

The reason I did this is that the `Autocomplete` component requires that the `value` is also part of `options`. If it's not, a warning (or error?) is raised. I'm not sure why this is not the case with your code, maybe because the text change triggers another search or something.

I think we should discuss these changes before we decide whether to implement them, go back, or use some other solution. But at the very least the bug will need to be fixed.

**Comment by djbusstop** [comment]
> The reason I did this is that the Autocomplete component requires that the value is also part of options. If it's not, a warning (or error?) is raised. I'm not sure why this is not the case with your code, maybe because the text change triggers another search or something.

To this point, I solved this in the `getOptionSelected` property of the Autocomplete, to allow it to accept the current user as well.

**Comment by djbusstop** [comment]
Ok I understand. To be honest, I think it's not a terrible time to make the person autoselect. But I get that it may be a bit soon. I'm very sorry that I made my changes in the merge commit, because of course it makes it much harder to revert my changes. I can revert to make the changes you introduced in the new form component.

**Review by djbusstop — APPROVED** [review]

**Pull request merged**